python/histone_utility.py

```diff
@@ -0,0 +1,355 @@
+                       
+    """
+
+    temp_df = df.copy()
+    
+    delete = temp_df[delete_column]
+    del temp_df[delete_column]
+    
+    x = temp_df.values.astype(float)
+    min_max_scaler = preprocessing.MinMaxScaler()
+    x_scaled = min_max_scaler.fit_transform(x)
+    df_normalized = pd.DataFrame(x_scaled)
+
+    df_normalized.insert(0, delete_column, delete, True)
+    df_normalized.columns = [delete_column] + cell_list
+    
+
+    # Fill in NaN values.
+    df_normalized = df_normalized.fillna(method='ffill')
+    return df_normalized
+
+
+def common(list_a,list_b):
+    """ This functions purpose is to find common elements of 
+    lists by converting each list to a set.
+    
+    The inputs of this function should both be lists. 
+    
+    INPUT
+        a: list which you wish to find common elements
+        b: list which you wish to find common elements
+        
+    OUTPUT
+        common: list containing common elements from a and b
+        
+    """
+    a_set = set(list_a)
+    b_set = set(list_b)
+    common = a_set.intersection(b_set) 
+    common = list(common)
+    return common 
+    
+def pearson_dfs(gene_list, histone_list, gene_data, histone_data, dtype):
+    """This functions purpose  is to compute the column-wise pearson 
+    correlation coefficient and p-value for each element in an array 
+    
+    The first dataframe will contain all the r-values between the two sets of data. 
+    The second will contain all the p-values for each r-value.
+    
+    This function is very similar to the Matrix function however it will 
+    create a dataframe with columns being histones and rows being genes. 
+   
+    This function is very specific. The dataframe created will contain histone
+    markers as the columns and gene names as the index. 
+    
+    This function assumes you already took the intersection between the two arrays. 
+    If you do not do that, then the matrix dimensions will not agree, 
+    resulting in an error.
+    
+    The format for inputting matrix should have columns 
+    correspond to the celllines.
+    
+    INPUT
+        gene_list: list of genes which correlation study will be performed
+        histone_list: list of histones which correlation study will be performed
+        gene_data: dataframe containing genes which you would like to compare
+        histone_data: dataframe containing histones which you would like to compare
+        dtype: 'numpy' or 'df', numpy returns a matrix, and df returns a dataframe
+            
+    OUTPUT
+    
+        Rmat: R-value matrix
+        Pmat: p-value matrix
+        Rdf: R-value dataframe
+        Pdf: p-value dataframe
+        
+    """
+    Rmat = (len(gene_list), len(histone_list))
+    Rmat = np.zeros(Rmat)
+    Pmat = Rmat[:]
+    i = 0 
+    
+    for gene in gene_list:
+        j = 0
+        for histone in histone_list:
+            gene_data.loc[gene] = gene_data.loc[gene].fillna(method='ffill')
```

Comment:
This part can actually be taken out of the for loop and can be vectorized.

Reply:
gene_data = gene_data.fillna(method='ffill') should work

Reply:
I think I fixed this one